Support Structured Outputs with enable_thinking=False in Qwen3

halaction opened the pull request:
## Motivation

Resolves https://github.com/sgl-project/sglang/issues/6675

The current `ReasoningGrammarBackend` doesn't support generation with `enable_thinking=False`. When this argument is passed in Qwen3 models no reasoning content is produced, and think tokens `<think>` and `</think>` are absent. The grammar backend assumes `ReasoningGrammarObject.is_in_reasoning=True` and expects `</think>` to start constrained decoding. However, the token never appears, hence structured outputs are not enforced.

## Modifications

Flag `enable_thinking` is passed per-request as `sampling_params["expect_thinking"]` to control `ReasoningGrammarObject.is_in_reasoning`. 

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

halaction committed:
fix: added starting think token processing to ReasonerGrammarBackend to properly handle enable_thinking

halaction committed:
fix: switch to passing enable_thinking per request through sampling_params

CatherineSue commented on python/sglang/srt/constrained/reasoner_grammar_backend.py:
```diff
@@ -76,15 +86,31 @@ def jump_and_retokenize(
     def _init_value_dispatch(
-        self, key: Tuple[str, str]
+        self,
+        key: Tuple[str, str, str],
```
Should it be 
```suggestion
        key: Tuple[str, str, bool],
```
?

CatherineSue commented on python/sglang/srt/managers/scheduler.py:
```diff
@@ -1037,14 +1037,24 @@ def handle_generate_request(
+                key_string = req.sampling_params.structural_tag
+
+            if self.server_args.reasoning_parser and self.tokenizer:
+                key = (key_type, key_string, req.sampling_params.expect_thinking)
```
Will there be a case when `req.sampling_params.expect_thinking` is `None` here? What happens then?

halaction replied:
Haven't found one so far. It is assigned in `adapter.py` (https://github.com/halaction/sglang/blob/e240c542491f495bac26d183a43fe801a6059a0a/python/sglang/srt/openai_api/adapter.py#L1179) and used in `scheduler.py` (https://github.com/halaction/sglang/blob/e240c542491f495bac26d183a43fe801a6059a0a/python/sglang/srt/managers/scheduler.py#L1055) after the same checks for constraints. Value for `expect_thinking` is received from `_get_enable_thinking_from_request`, which is guaranteed to return `True` or `False`.

CatherineSue reviewed (commented):
Overall LGTM. Thanks for fixing this bug. 
Could you add a UT if possible?

halaction committed:
Apply suggestions from code review

Co-authored-by: Chang Su <csu272@usc.edu>

zhyncs committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs

zhyncs commented:
Hi @halaction May you update this PR

zhyncs committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs

halaction committed:
fix: pass enable_thinking to sampling_params unconditionally and some general refactoring

halaction committed:
feat: added tests for enable_thinking with structured outputs

halaction commented:
Hi, I added two unit tests with XGrammar and JSON but I wonder if all backends and formats should be tested with enable_reasoning. Also not sure if it needs tests with tools as another kind of constrained decoding. Let me know what is your policy, I can add more cases

halaction committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs

JustinTong0323 committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs

JustinTong0323 commented:
There are some errors in CI, plz check.
<img width="957" alt="image" src="https://github.com/user-attachments/assets/f05d489d-1fd8-4f38-bc26-e937e18383b5" />
(this is in unittest7)

halaction committed:
fix: resolved bugs with response format and json content in enable_thinking tests

halaction committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs

halaction commented:
Fixed bugs in test_enable_thinking. You can run tests again

halaction committed:
Merge remote-tracking branch 'origin/main' into hotfix/qwen3-structured-outputs

zhyncs committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs

halaction committed:
fix: ensure definite enable_thinking in scheduler

halaction committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs

halaction committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs

yizhang2077 commented on python/sglang/srt/constrained/reasoner_grammar_backend.py:
```diff
@@ -76,15 +86,31 @@ def jump_and_retokenize(
-        return ReasonerGrammarObject(ret, self.think_end_id)
+        return ReasonerGrammarObject(ret, self.think_end_id, enable_thinking)
+
+    def get_cached_or_future_value(
```
this is the same as `get_cached_or_future_value` in `BaseGrammarBackend`?

halaction replied:
It is just a typing overload, the only difference is the signature of `key`

halaction committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs

halaction commented:
Except for linting test fails seems to be unrelated to the bug. Ready to merge?

frontword commented:
This PR can fix the issue when stream=False, but when stream=True, the output is not structured json format

JustinTong0323 commented:
> > This PR can fix the issue when stream=False, but when stream=True, the output is not structured json format
> 
> @frontword Sorry for the late reply. I tested it locally and streaming provides the same answers as non-streaming with the current fix. Could you provide the conflicting results or point to the corresponding tests? I will add tests for streaming ASAIC to this PR

Hi, Thanks for the contribution! Could you please resolve the conflicts so we could speed up merging this PR? Sorry for missing catch up for a while.

halaction committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs

halaction committed:
feat: added tests for streaming structured outputs depending on enable_thinking

halaction committed:
fix: comply with new behaviour in _get_enable_thinking_from_request and force reasoning for non-hybrid models

halaction committed:
fix: rollback to enable_thinking and specify is_hybrid in ReasoningParser

halaction committed:
fix: interpret unspecified enable_thinking as False

halaction committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs

halaction committed:
chore: pre-commit

halaction commented:
I updated the PR to accommodate hybrid and non-hybrid models as defined in reasoning detectors, added tests for streaming and resolved the conflicts. Ready for testing

halaction committed:
Merge branch 'main' into hotfix/qwen3-structured-outputs
